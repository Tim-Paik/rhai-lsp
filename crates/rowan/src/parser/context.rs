//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
