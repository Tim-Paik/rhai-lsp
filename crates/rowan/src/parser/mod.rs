//! This module contains all parsing-related tools including a
//! recursive-descent parser.

use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

/// A flexible parser.
/// 
/// Parsing happens via given parser functions (also found in [`parsers`]).
/// 
/// # Example Uses
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// // Parse some Rhai code (file).
/// let parser = Parser::new(r#"print("hello rhai!")"#);
/// let parse = parser.parse();
/// 
/// assert!(parse.errors.is_empty());
/// ```
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// use rhai_rowan::parser::parsers::parse_expr;
///
/// // Parse an expression explicitly.
/// let mut parser = Parser::new(r#"print("hello rhai!")"#);
/// parser.execute(parse_expr);
/// let parse = parser.finish();
/// 
/// assert!(parse.errors.is_empty());
/// ```
pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    /// Create a new parser with the given source.
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    /// Finish parsing.
    /// 
    /// # Panics
    /// 
    /// If no parser function was called,
    /// or a parser was left in an invalid state (bug).
    #[must_use]
    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    /// Execute a parser function for the given parser.
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context);
    }
}

/// The result of parsing, containing all errors
/// and the final green tree.
#[derive(Debug, Clone)]
pub struct Parse {
    /// Parse errors.
    pub errors: Vec<ParseError>,
    /// Parsed green tree.
    pub green: GreenNode,
}

impl Parse {
    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn clone_syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }

    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

/// A parse (syntax) error.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    /// The span of the error in the parsed source.
    pub range: rowan::TextRange,
    /// Error kind.
    pub kind: ParseErrorKind,
}

impl ParseError {
    fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

/// All the non-fatal parsing errors that can occur.
#[derive(Debug, PartialEq, Eq, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .collect::<Vec<String>>().join(",")
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
