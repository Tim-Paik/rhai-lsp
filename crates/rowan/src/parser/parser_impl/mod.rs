#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
