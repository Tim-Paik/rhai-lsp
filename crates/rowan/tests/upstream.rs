use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../rhai/scripts/array.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/assignment.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/comments.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/fibonacci.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl4.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/loop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/mat_mul.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/module.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/oop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/primes.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/speed_test.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/string.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/strings_map.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/switch.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/while.rhai"))]
fn parse_rhai_upstream(src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty());
    // TODO(tests)
    // insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
