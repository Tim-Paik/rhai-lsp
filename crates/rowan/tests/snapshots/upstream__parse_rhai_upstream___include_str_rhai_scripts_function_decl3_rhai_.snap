---
source: crates/rowan/tests/upstream.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
RHAI@0..233
  COMMENT_LINE@0..67 "// This script define ..."
  WHITESPACE@67..69 "\n\n"
  STMT@69..84
    ITEM@69..83
      EXPR@69..83
        EXPR_CONST@69..83
          KW_CONST@69..74 "const"
          WHITESPACE@74..75 " "
          IDENT@75..78 "KEY"
          WHITESPACE@78..79 " "
          OP_ASSIGN@79..80 "="
          EXPR@80..83
            WHITESPACE@80..81 " "
            EXPR_LIT@81..83
              LIT@81..83
                LIT_INT@81..83 "38"
    PUNCT_SEMI@83..84 ";"
  WHITESPACE@84..86 "\n\n"
  STMT@86..154
    ITEM@86..152
      EXPR@86..152
        EXPR_FN@86..152
          KW_FN@86..88 "fn"
          WHITESPACE@88..89 " "
          IDENT@89..90 "f"
          PARAM_LIST@90..108
            PUNCT_PAREN_START@90..91 "("
            PARAM@91..92
              IDENT@91..92 "a"
            PUNCT_COMMA@92..93 ","
            WHITESPACE@93..94 " "
            PARAM@94..95
              IDENT@94..95 "b"
            PUNCT_COMMA@95..96 ","
            WHITESPACE@96..97 " "
            PARAM@97..98
              IDENT@97..98 "c"
            PUNCT_COMMA@98..99 ","
            WHITESPACE@99..100 " "
            PARAM@100..101
              IDENT@100..101 "d"
            PUNCT_COMMA@101..102 ","
            WHITESPACE@102..103 " "
            PARAM@103..104
              IDENT@103..104 "e"
            PUNCT_COMMA@104..105 ","
            WHITESPACE@105..106 " "
            PARAM@106..107
              IDENT@106..107 "f"
            PUNCT_PAREN_END@107..108 ")"
          EXPR_BLOCK@108..152
            WHITESPACE@108..109 " "
            PUNCT_BRACE_START@109..110 "{"
            WHITESPACE@110..115 "\n    "
            STMT@115..151
              ITEM@115..151
                EXPR@115..151
                  EXPR_BINARY@115..151
                    EXPR@115..137
                      EXPR_BINARY@115..137
                        EXPR@115..133
                          EXPR_BINARY@115..133
                            EXPR@115..125
                              EXPR_BINARY@115..125
                                EXPR@115..117
                                  EXPR_IDENT@115..117
                                    IDENT@115..116 "a"
                                    WHITESPACE@116..117 " "
                                OP_SUB@117..118 "-"
                                EXPR@118..125
                                  EXPR_BINARY@118..125
                                    EXPR@118..121
                                      WHITESPACE@118..119 " "
                                      EXPR_IDENT@119..121
                                        IDENT@119..120 "b"
                                        WHITESPACE@120..121 " "
                                    OP_MUL@121..122 "*"
                                    EXPR@122..125
                                      WHITESPACE@122..123 " "
                                      EXPR_IDENT@123..125
                                        IDENT@123..124 "c"
                                        WHITESPACE@124..125 " "
                            OP_SUB@125..126 "-"
                            EXPR@126..133
                              EXPR_BINARY@126..133
                                EXPR@126..129
                                  WHITESPACE@126..127 " "
                                  EXPR_IDENT@127..129
                                    IDENT@127..128 "d"
                                    WHITESPACE@128..129 " "
                                OP_MUL@129..130 "*"
                                EXPR@130..133
                                  WHITESPACE@130..131 " "
                                  EXPR_IDENT@131..133
                                    IDENT@131..132 "e"
                                    WHITESPACE@132..133 " "
                        OP_SUB@133..134 "-"
                        EXPR@134..137
                          WHITESPACE@134..135 " "
                          EXPR_IDENT@135..137
                            IDENT@135..136 "f"
                            WHITESPACE@136..137 " "
                    OP_ADD@137..138 "+"
                    EXPR@138..151
                      WHITESPACE@138..139 " "
                      EXPR_PATH@139..151
                        PATH@139..151
                          IDENT@139..145 "global"
                          PUNCT_COLON2@145..147 "::"
                          IDENT@147..150 "KEY"
                          WHITESPACE@150..151 "\n"
            PUNCT_BRACE_END@151..152 "}"
    WHITESPACE@152..154 "\n\n"
  STMT@154..190
    ITEM@154..189
      EXPR@154..189
        EXPR_LET@154..189
          KW_LET@154..157 "let"
          WHITESPACE@157..158 " "
          IDENT@158..164 "result"
          WHITESPACE@164..165 " "
          OP_ASSIGN@165..166 "="
          EXPR@166..189
            EXPR_CALL@166..189
              EXPR@166..168
                WHITESPACE@166..167 " "
                EXPR_IDENT@167..168
                  IDENT@167..168 "f"
              ARG_LIST@168..189
                PUNCT_PAREN_START@168..169 "("
                EXPR@169..172
                  EXPR_LIT@169..172
                    LIT@169..172
                      LIT_INT@169..172 "100"
                PUNCT_COMMA@172..173 ","
                WHITESPACE@173..174 " "
                EXPR@174..175
                  EXPR_LIT@174..175
                    LIT@174..175
                      LIT_INT@174..175 "5"
                PUNCT_COMMA@175..176 ","
                WHITESPACE@176..177 " "
                EXPR@177..178
                  EXPR_LIT@177..178
                    LIT@177..178
                      LIT_INT@177..178 "2"
                PUNCT_COMMA@178..179 ","
                WHITESPACE@179..180 " "
                EXPR@180..181
                  EXPR_LIT@180..181
                    LIT@180..181
                      LIT_INT@180..181 "9"
                PUNCT_COMMA@181..182 ","
                WHITESPACE@182..183 " "
                EXPR@183..184
                  EXPR_LIT@183..184
                    LIT@183..184
                      LIT_INT@183..184 "6"
                PUNCT_COMMA@184..185 ","
                WHITESPACE@185..186 " "
                EXPR@186..188
                  EXPR_LIT@186..188
                    LIT@186..188
                      LIT_INT@186..188 "32"
                PUNCT_PAREN_END@188..189 ")"
    PUNCT_SEMI@189..190 ";"
  WHITESPACE@190..192 "\n\n"
  STMT@192..232
    ITEM@192..231
      EXPR@192..231
        EXPR_CALL@192..231
          EXPR@192..197
            EXPR_IDENT@192..197
              IDENT@192..197 "print"
          ARG_LIST@197..231
            PUNCT_PAREN_START@197..198 "("
            EXPR@198..230
              EXPR_LIT@198..230
                LIT@198..230
                  LIT_STR@198..230 "`result should be 42: ..."
            PUNCT_PAREN_END@230..231 ")"
    PUNCT_SEMI@231..232 ";"
  WHITESPACE@232..233 "\n"

